File: backend/src/services/shoppingListGenerator.js
const { getDb } = require("../db");

// Meal cutoff hours (same as frontend)
const MEAL_CUTOFF = { breakfast: 10, lunch: 14, dinner: 20 };

/**
 * Check if a meal slot is in the past based on weekStart, dayIndex, mealType.
 */
function isSlotExpired(weekStart, dayIndex, mealType) {
  const now = new Date();
  const monday = new Date(weekStart + "T00:00:00");
  const slotDate = new Date(monday);
  slotDate.setDate(monday.getDate() + dayIndex);
  slotDate.setHours(0, 0, 0, 0);

  const today = new Date();
  today.setHours(0, 0, 0, 0);

  if (slotDate < today) return true;
  if (slotDate.getTime() === today.getTime()) {
    const cutoff = MEAL_CUTOFF[mealType];
    if (cutoff && now.getHours() >= cutoff) return true;
  }
  return false;
}

/**
 * Get aggregated shopping list filtered by selections.
 * @param {string} userId
 * @param {string} weekStart - YYYY-MM-DD
 * @param {Array<{dayIndex: number, mealType: string}>|null} selections - if null, returns empty list
 */
function getShoppingList(userId, weekStart, selections) {
  const db = getDb();
  const ws = weekStart || getCurrentWeekStart();

  // If no selections provided, return empty
  if (!selections || selections.length === 0) {
    return [];
  }

  // Build filter for selected slots
  const entries = db.prepare(`
    SELECT mp.day_index, mp.meal_type, mp.servings, mp.custom_ingredients, d.ingredients
    FROM meal_plans mp
    JOIN dishes d ON d.id = mp.dish_id
    WHERE mp.user_id = ? AND mp.week_start = ?
  `).all(userId, ws);

  // Create a set of selected slot keys for fast lookup
  const selectedKeys = new Set(
    selections.map(s => `${s.dayIndex}-${s.mealType}`)
  );

  const list = {};

  for (const entry of entries) {
    const key = `${entry.day_index}-${entry.meal_type}`;
    if (!selectedKeys.has(key)) continue;

    const dishIngredients = JSON.parse(entry.ingredients);
    const customIngs = entry.custom_ingredients ? JSON.parse(entry.custom_ingredients) : null;

    const ings = customIngs || Object.fromEntries(
      Object.entries(dishIngredients).map(([k, v]) => [k, v * entry.servings])
    );

    for (const [ing, amount] of Object.entries(ings)) {
      list[ing] = (list[ing] || 0) + amount;
    }
  }

  return Object.entries(list)
    .sort((a, b) => a[0].localeCompare(b[0]))
    .map(([name, grams]) => ({ name, grams: Math.round(grams) }));
}

function getCurrentWeekStart() {
  const d = new Date();
  const day = d.getDay();
  const diff = day === 0 ? -6 : 1 - day;
  d.setDate(d.getDate() + diff);
  return d.toISOString().split("T")[0];
}

module.exports = { getShoppingList, isSlotExpired };

File: backend/services/shopping_list_generator.py
from __future__ import annotations

from datetime import date, datetime, timedelta
import sqlite3

from ..constants import MEAL_CUTOFF
from ..utils import get_current_week_start, parse_json


def is_slot_expired(week_start: str, day_index: int, meal_type: str) -> bool:
    now = datetime.now()
    monday = datetime.fromisoformat(f"{week_start}T00:00:00")
    slot_date = monday + timedelta(days=int(day_index))

    today = date.today()
    if slot_date.date() < today:
        return True
    if slot_date.date() == today:
        cutoff = MEAL_CUTOFF.get(meal_type)
        if cutoff is not None and now.hour >= cutoff:
            return True
    return False


def get_shopping_list(
    conn: sqlite3.Connection,
    user_id: str,
    week_start: str | None,
    selections: list[dict[str, int | str]] | None,
) -> list[dict[str, int | str]]:
    ws = week_start or get_current_week_start()

    if not selections:
        return []

    rows = conn.execute(
        """
        SELECT mp.day_index, mp.meal_type, mp.servings, mp.custom_ingredients, d.ingredients
        FROM meal_plans mp
        JOIN dishes d ON d.id = mp.dish_id
        WHERE mp.user_id = ? AND mp.week_start = ?
        """,
        (user_id, ws),
    ).fetchall()

    selected_keys = {f"{s['dayIndex']}-{s['mealType']}" for s in selections}

    totals: dict[str, float] = {}
    for row in rows:
        key = f"{row['day_index']}-{row['meal_type']}"
        if key not in selected_keys:
            continue

        dish_ingredients = parse_json(row["ingredients"], {})
        custom = parse_json(row["custom_ingredients"], None)
        ingredients = custom or {
            k: float(v) * float(row["servings"]) for k, v in dish_ingredients.items()
        }

        for ingredient, amount in ingredients.items():
            totals[ingredient] = totals.get(ingredient, 0) + float(amount)

    return [
        {"name": ingredient, "grams": int(round(grams))}
        for ingredient, grams in sorted(totals.items(), key=lambda kv: kv[0])
    ]
